client/auth.go

```diff
@@ -0,0 +1,394 @@
+}
+
+// Deletes role.
+func (c *authClient) RoleDelete(request *xlineapi.AuthRoleDeleteRequest) (*xlineapi.AuthRoleDeleteResponse, error) {
+	requestWithToken := xlineapi.RequestWithToken{
+		Token: &c.token,
+		RequestWrapper: &xlineapi.RequestWithToken_AuthRoleDeleteRequest{
+			AuthRoleDeleteRequest: request,
+		},
+	}
+	resp, err := c.handleReq(&requestWithToken, false)
+	if err != nil {
+		return nil, err
+	}
+	res := resp.CommandResp.GetAuthRoleDeleteResponse()
+	res.Header.Revision = resp.SyncResp.Revision
+	return res, err
+}
+
+// Grants role permission
+func (c *authClient) RoleGrantPermission(request *xlineapi.AuthRoleGrantPermissionRequest) (*xlineapi.AuthRoleGrantPermissionResponse, error) {
+	if request.Perm == nil {
+		return nil, errors.New("Permission not given")
+	}
+
+	requestWithToken := xlineapi.RequestWithToken{
+		Token: &c.token,
+		RequestWrapper: &xlineapi.RequestWithToken_AuthRoleGrantPermissionRequest{
+			AuthRoleGrantPermissionRequest: request,
+		},
+	}
+	resp, err := c.handleReq(&requestWithToken, false)
+	if err != nil {
+		return nil, err
+	}
+	res := resp.CommandResp.GetAuthRoleGrantPermissionResponse()
+	res.Header.Revision = resp.SyncResp.Revision
+	return res, err
+}
+
+// Revokes role permission
+func (c *authClient) RoleRevokePermission(request *xlineapi.AuthRoleRevokePermissionRequest) (*xlineapi.AuthRoleRevokePermissionResponse, error) {
+	requestWithToken := xlineapi.RequestWithToken{
+		Token: &c.token,
+		RequestWrapper: &xlineapi.RequestWithToken_AuthRoleRevokePermissionRequest{
+			AuthRoleRevokePermissionRequest: request,
+		},
+	}
+	resp, err := c.handleReq(&requestWithToken, false)
+	if err != nil {
+		return nil, err
+	}
+	res := resp.CommandResp.GetAuthRoleRevokePermissionResponse()
+	res.Header.Revision = resp.SyncResp.Revision
+	return res, err
+}
+
+// Send request using fast path
+func (c authClient) handleReq(req *xlineapi.RequestWithToken, useFastPath bool) (*ProposeResponse, error) {
+	proposeId := c.generateProposeId()
+	cmd := xlineapi.Command{Request: req, ProposeId: proposeId}
+
+	if useFastPath {
+		res, err := c.curpClient.propose(&cmd, true)
+		return res, err
+	} else {
+		res, err := c.curpClient.propose(&cmd, false)
+		if err != nil {
+			return res, err
+		} else {
+			if res != nil && res.SyncResp == nil {
+				panic("syncResp is always Some when useFastPath is false")
+			}
+			return res, err
+		}
+	}
+}
+
+// Generate hash of the password
+func (c authClient) hashPassword(password []byte) string {
+	const (
+		pbkdf2Iter   = 10000
+		pbkdf2KeyLen = 32
+	)
+
+	salt := make([]byte, 16)
+	_, err := rand.Read(salt)
+	// This doesn't seems to be fallible
+	if err != nil {
+		panic(err)
+	}
+
+	hashed := pbkdf2.Key(password, salt, pbkdf2Iter, pbkdf2KeyLen, sha256.New)
+
+	_salt := base64.RawStdEncoding.EncodeToString(salt)
+	_hashed := base64.RawStdEncoding.EncodeToString(hashed)
+
+	hash := fmt.Sprintf("$pbkdf2-sha256$i=%d,l=%d$%s$%s", pbkdf2Iter, pbkdf2KeyLen, _salt, _hashed)
```

Comment:
`PHC` is a specification. Is there any package in go could handle with it?

Reply:
Go language does not seem to have a mature library dedicated to generating PHC format password hashes.